### ISSUE TITLE
Support fragment calculations in some way

### ISSUE THREAD

**Issue by langner** [cclib/cclib#131]
Calculations based on fragments are getting more popular. Currently in cclib our approach is to parse the supermolecule or largest/complete system in a job. Since we mightwant  to change the API for other reasons (#99), we should try at the same time to accommodate fragments in some way.

Related: #116

**Comment by Clyde-fare** [comment]
For fragments or ONIOM what about modifying the parsing code in logfileparser such that we have something like:

        frag_ids = self.get_frag_ids()
        self.frag_parsers = {frag_id: copy.deepcopy(self) for frag_id in frag_ids}

        for line in inputfile:

            self.updateprogress(inputfile, "Unsupported information", cupdate)
            
            if self.frag_id:
                current_parser = self.frag_parsers[self.frag_id]
                current_parser.extract(inputfile, line)
                self.frag_id = current_parser.frag_id
            else:
                self.extract(inputfile, line)

I.e. we have separate parser objects associated with each fragment (and the master associated with all non-fragment information) the parsers then have sections that  look for lines indicating the start or end a particular fragment section and update their fragment_id attribute to indicate to either empty if we are leaving the fragment info or to the relevant id if we are changing to a new fragment.

Then at we would extract the data from the fragment parsing objects to ccdata objects and attach them to the master ccdata object.

**Comment by langner** [comment]
@Clyde-fare I'm not sure I understand what you mean. What is the point of copying the parser for each fragment, if it is the same? 

(linking to #161 for reference.)

**Comment by Clyde-fare** [comment]
I was thinking that following an initial scan determining how many fragments are involved

A parser for each fragment would be created by copying the main parser.

Then for each iteration of the master loop present in logfileparser a fields frag_id would be used to determine which fragment parser should be used  to parse the next inputline.

I had a go at getting it to work see my [branch](https://github.com/Clyde-fare/cclib/tree/fragment_parsing) but haven't quite got it working yet. 

I though it would be a reasonable idea because data corresponding to the fragments would be naturally separated inside fragments parser and the process should work well if the output appearing in the log file that corresponds to a particular fragment has an identical layout to equivalent output in a normal calculation.

However having tried it out I think there's probably a simpler solution. My main problems were:

* It complicates the master loop present in logfileparser.
* Copying the parsers has to be taken care of as the Logfile object they contain is not copyable.
* Sychronisation between the fragment parser objects is not entirely straightforward. 

The latter case was gaving me headaches as I wasn't sure how much information should be shared across the different fragment parsers and how exactly to do the sharing without over-complicating the code.

**Comment by langner** [comment]
> The latter case was giving me headaches as I wasn't sure how much information should be shared across the different fragment parsers and how exactly to do the sharing without over-complicating the code.

Potential complications do seem to be the main danger in your approach. I think that a parser is inherently tied with the file type or program, so I would hesitate very much before multiplying parsers. The idea of logfile->parser->data is simple and easy to work with, and I fear changing that flow would create more problems than advantages.

**Comment by langner** [comment]
> The idea of logfile->parser->data is simple and easy to work with, and I fear changing that flow too much would create more problems than advantages.

To expand on this flow paradigm, your approach seems to be logfile->[parsers]->[data], which means there are two levels at which things can diverge. In contrast, if we stick with a single parser per logfile type and user logfile->parser->[data], we only need to worry about when to create a new data object.

**Comment by langner** [comment]
Create https://github.com/cclib/cclib/tree/fragments for experimenting with fragments.

**Comment by Clyde-fare** [comment]
Hi thanks for the explanation it makes a lot of sense.

So then would the approach be to create a fragments objects inside the parser object which would contain a list or dictionary of fragment data objects?

What about refactoring the calculator so instead of having lines like:

     if not hasattr(self, "scftargets"):
                    self.scftargets = []
    ...
    self.scftargets.append(blah)

Instead we have a line like

    self.save_data('scftargets', blah)

Where the save_data method takes care of figuring out where to put the new scftarget (and creating the relevant field if it doesn't already exist)

**Comment by langner** [comment]
> Hi thanks for the explanation it makes a lot of sense.
> 
> So then would the approach be to create a fragments objects inside the parser object which would contain a list or dictionary of fragment data objects?

I'm not sure, I've only begun to think about this again in the past few days. I looked through some ONIOM outputs, and its not as simple as I thought previously. I'm beginning to tend towards your solution in the case of ONIOM, that is to include an attribute(s) to hold the extra ONIOM data.

> What about refactoring the calculator so instead of having lines like:
> 
>  if not hasattr(self, "scftargets"):
>                 self.scftargets = []
> ...
> self.scftargets.append(blah)
> Instead we have a line like
> 
> self.save_data('scftargets', blah)
> Where the save_data method takes care of figuring out where to put the new scftarget (and creating the relevant field if it doesn't already exist)

We've gone somewhat in that direction of abstraction by customizing the setter on Logfile and creating a `set_attribute` method for routine checks, which is used now by many parsers:
https://github.com/cclib/cclib/blob/master/src/cclib/parser/logfileparser.py#L341

It would make sense to go further in that direction, although often the parsing is quite context specific. Do you have any particular code fragment in mind?

**Comment by Clyde-fare** [comment]
I was thinking of the scf values reported in ONIOM involving two ab-initio methods or an ab-initio and a semi empirical method.  At the moment the values get muddled together into the .scfvalues field. I guess the same phenomena is happening when you try and use something like GaussSum to monitor geometry optimisation of oniom - it assumes each instance of SCF=Done is an energy associated with a particular geometry which means you get crazy oscillating plots for oniom calculation involving sem-empirical low level methods.

More generally I was thinking that about fragment calculations where the output associated with a fragment looks just like a normal non-fragment output, but perhaps that's not as big a use-case as I imagined.

**Comment by langner** [comment]
Another idea: **add an attribute for fragments that is a list of ccData objects**

Pros:
1) Don't need to change API too much
2) Presents a natural hierarchy, since fragments "belong" to the parent ccData object
3) Allows for easy, recursive export to JSON or whatever
4) Can be parsed equentially along with the parent model, in little chunks or however a particular program prints
5) Can be used for fragments, BSSE, ONIOM, ADF fragments and probably more

Cons: ?

**Comment by Clyde-fare** [comment]
As far as I understand it at the moment the ccData object ins't populated until after the file has been parsed.

I guess a possible downside is that the master file and the fragments would be treated differently in the sense that parsed data for fragments would be placed directly into a ccData object as parsing progressed where as parsed data from the main body is stored in the internal dictionary of the parser and only placed into the ccData object that end.

Could perhaps use a list dictionaries instead of a list of ccData objects, and then pull those dictionaries into ccData objects after the file has been parsed?

**Comment by langner** [comment]
You're right... but I don't see that as a major obstacle. The choice to store data temporarily in the parser's `__dict__` did not have any strong motivation. In fact, many releases ago there was no ccData object, so this was just the simplest thing to do. I would not hesitate to change that if there was a good reason to do so.

Cons:
1) Might require reworking the logic of parsing a little bit.

**Comment by berquist** [comment]
I've started thinking about this again, since I'd really like to have it.

Of the 3 points that @Clyde-fare brings up, I might have a solution to

> Sychronisation between the fragment parser objects is not entirely straightforward.

which would be to avoid doing this mostly until the very end. It would complicate `logfileparser` and other high-level code at the expense of not complicating individual parsers, and potentially even simplifying them. Actually, this is identical to what he suggests but moves the "spawning" up another level.

To use @langner's notation, my idea would be this workflow:

```
logfile->[logfiles]
for each logfile in [logfiles]:
    logfile->parser->data
```

The basic idea is that during a "pre-attribute-parsing" run, the boundaries of each fragment are determined and stored. These boundaries are used to split up the logfile into multiple temporary files kept in memory. For each of these, a logfile of the appropriate type is spawned for only that file's contents, and `parse()` is called.

I fear that spawning ccdata objects and keeping track of them within each parser is dangerous in terms of the bugs it could introduce and it would duplicate a lot of shared logic for doing this spawning and synchronization.

I also think that
> More generally I was thinking that about fragment calculations where the output associated with a fragment looks just like a normal non-fragment output, but perhaps that's not as big a use-case as I imagined.

is actually very common. ONIOM can probably deviate, but BSSE-correction, many-body expansion, and I think ADF fragments all follow this model. There are always going to be exceptions, and these would probably have to be handled in individual parsers no matter what.

Pros:

1. At a certain level, the flow of `logfile->parser->data` is maintained.
2. Reduces duplication of complicated shared "fragment management" logic in individual parsers.
3. Potentially simplifies current parsers by removing some "supersystem"-parsing only logic.
4. Treats all fragments on an equal footing within parsers (with the potential exception of the master).
5. Parsers can be spawned and run asynchronously.

Cons

1. There might be "core" attributes (`nbasis, nmo, ...`) that don't appear in the fragments; this is deviation from that "normal" output displayed above. There would have to be a synchronization call that fixes this.
2. Asynchronous code can be very hard to debug. This could and should initially be sequential.
3. Potentially very significant re-write of the high-level code.
